torch/backends/_nnapi/serializer.py

```diff
@@ -0,0 +1,1131 @@
+            op_id = self.add_tensor_operand_for_input(input_value, input_tensor)
+            inp_dim_orders.append(self.operands[op_id].dim_order.value)
+
+        for idx, node in enumerate(model.graph.nodes()):
+            LOG.debug("Processing node #%d: %r", idx, node)
+            self.add_node(node)
+
+        retn = model.graph.return_node()
+        assert retn.inputsSize() == 1
+        assert retn.outputsSize() == 0
+        # TODO: Make outputs a local variable? 
+        # TODO: Handle tuple-of-tensor return
+        for idx in range(1):
+            op_id = self.jitval_operand_map[retn.inputsAt(0)]
+            self.outputs.append(op_id)
+            out_dim_orders.append(self.operands[op_id].dim_order.value)
+
+        model = []
+
+        version = 1
+        header = struct.pack(
+                "iiiiii",
+                version,
+                len(self.operands),
+                len(self.values),
+                len(self.operations),
+                len(self.inputs),
+                len(self.outputs),
+                )
+        model.append(header)
+
+        serialized_values, serialized_value_data = self.serialize_values()
+
+        model.extend(struct.pack("iifi", t, len(d), s, z) for (t,d,_m,s,z) in self.operands)
+        model.extend(serialized_values)
+        model.extend(struct.pack("iii", *x) for x in self.operations)
+        model.extend(self.serialize_ints(fix_shape(dims, mf)) for (_, dims, mf, _, _) in self.operands)
+        model.extend(serialized_value_data)
+        model.append(self.serialize_ints(self.operation_args))
+        model.append(self.serialize_ints(self.inputs))
+        model.append(self.serialize_ints(self.outputs))
+
+        #return (b"".join(model), self.used_weight_tensor_names)
+        return (b"".join(model), self.used_weights, inp_dim_orders, out_dim_orders)
+
+
+    def serialize_values(self):
+        serialized_values = []
+        serialized_value_data = []
+        assert len(self.values) == len(self.value_data)
+        for ((op_index, source_type), data) in zip(self.values, self.value_data):
+            source_length = len(data)
+
+            # Pad with 0 bytes out to a multiple of 4 for alignment.
+            physical_length = ((source_length - 1) | 0x3) + 1
+            padded_data = data + (b"\0" * (physical_length - source_length))
+
+            serialized_values.append(struct.pack("iii", op_index, source_type, source_length))
+            serialized_value_data.append(padded_data)
+
+        return serialized_values, serialized_value_data
+
+
+    @staticmethod
+    def serialize_ints(ints):
+        return struct.pack("i" * len(ints), *ints)
+
+
+    ADDER_MAP = {
+            "prim::GetAttr": lambda self, node:
+                self.add_getattr(node),
+            "prim::Constant": lambda self, node:
+                self.add_constant_node(node),
+            "prim::ListConstruct": lambda self, node:
```

Comment:
Could nnapi has corresponding prim::DictConstruct, TupleConstruct, etc? Do you have any plan to handle those nodes?

Reply:
NNAPI has no concepts of lists and dicts.  It only operates on tensors and scalars: https://android.googlesource.com/platform/frameworks/ml/+/refs/tags/android-11.0.0_r8/nn/runtime/include/NeuralNetworks.h#68

Reply:
Like how you handled prim::ListConstruct, we may also write code to handle DictConstruct and TupleConstruct from torchscript side?

Reply:
Oh, yes, I suppose so.